packages/server/src/health/service.ts

```diff
@@ -0,0 +1,147 @@
+import { CachingService } from '../caching/service'
+import { ChannelService } from '../channels/service'
+import { ClientService } from '../clients/service'
+import { ConduitService } from '../conduits/service'
+import { ConfigService } from '../config/service'
+import { DatabaseService } from '../database/service'
+import { InstanceService } from '../instances/service'
+import { WebhookService } from '../webhooks/service'
+import { HealthTable } from './table'
+import { HealthEvent, HealthEventType, HealthReport, HealthReportEvent } from './types'
+import { HealthWatcher } from './watcher'
+
+export class HealthService extends Service {
+  private table: HealthTable
+  private cache!: ServerCache<uuid, HealthReport>
+  private watcher: HealthWatcher
+
+  constructor(
+    private configService: ConfigService,
+    private db: DatabaseService,
+    private cachingService: CachingService,
+    private channelService: ChannelService,
+    private clientService: ClientService,
+    private webhookService: WebhookService,
+    private conduitService: ConduitService,
+    private instanceService: InstanceService
+  ) {
+    super()
+    this.table = new HealthTable()
+    this.watcher = new HealthWatcher(this.conduitService, this.instanceService, this)
+  }
+
+  async setup() {
+    this.cache = await this.cachingService.newServerCache('cache_health_reports')
+
+    await this.db.registerTable(this.table)
+
+    await this.watcher.setup()
+  }
+
+  async register(conduitId: uuid, type: HealthEventType, data: any = undefined) {
+    const event: HealthEvent = {
+      id: uuidv4(),
+      conduitId,
+      time: new Date(),
+      type,
+      data
+    }
+
+    const conduit = await this.conduitService.get(conduitId)
+    const client = await this.clientService.getByProviderId(conduit!.providerId)
+    if (client) {
+      this.cache.del(client.id, true)
+
+      const channel = this.channelService.getById(conduit!.channelId)
+      const post = {
+        type: 'health',
+        client: { id: client.id },
+        channel: { id: channel.id, name: channel.name },
+        event: this.makeReadable(event)
+      }
+
+      // TODO: this code to call webhook is copy pasted. It should be refactored somewhere else
+      if (process.env.SPINNED) {
+        await this.callWebhook(process.env.SPINNED_URL!, post)
+      } else {
+        const webhooks = await this.webhookService.list(client.id)
+
+        for (const webhook of webhooks) {
+          await this.callWebhook(webhook.url, post)
+        }
+      }
+    }
+
+    await this.query().insert(this.serialize(event))
+  }
+
+  // TODO: move this somewhere else
+  private async callWebhook(url: string, data: any) {
+    const password = process.env.INTERNAL_PASSWORD || this.configService.current.security?.password
+
+    try {
+      await axios.post(url, data, { headers: { password } })
+    } catch (e) {
+      // TODO: maybe we should retry if this call fails
```

Comment:
Totally agree. We might have to look at something like an incremental backoff strategy (as most channels do).

Reply:
Yeah I think a CallbackService or something could take care of registering important webhook calls and retrying them if they fail. I'd avoid putting that logic in WebhookService considering this service already takes care of the webhook tables. The CallbackService might have a table to keep track of failed calls to webhooks.

We won't enable this feature for messages for performance reason, but we'll need something like that for more "administrative" webhook such as health events.